navigator/backend/src/main/scala/com/digitalasset/navigator/store/platform/PlatformStore.scala

```diff
@@ -160,36 +162,92 @@ class PlatformStore(
+          .map { case (users, _) =>
+            users.toList
+          }
+
+      hasUserManagement
+        .flatMap {
+          case true =>
+            listAllUsers(state.ledgerClient)
+              .map(details => UpdatedUsers(details): Any)
+          //                                       ^^^^^
+          // make wartremover happy about the lub of UpdatedUsers and UpdatedParties
+          case false =>
+            state.ledgerClient.partyManagementClient
+              .listKnownParties()
+              .map(details => UpdatedParties(details): Any)
+        }
         .pipeTo(self)
+
       ()
+
+    case UpdatedUsers(users) =>
+      // Note: you cannot log in as a user without a primary party
+      val usersWithPrimaryParties = users.flatMap { user =>
+        user.primaryParty.map(p => (user.id, p))
+      }
+
+      users.filter(_.primaryParty.isEmpty) match {
+        case usersWithoutPrimaryParty if !usersWithoutPrimaryParty.isEmpty =>
+          log.warning(
+            s"Users without primary party (counted ${usersWithoutPrimaryParty.length})" +
+              s" cannot be used for login ${usersWithoutPrimaryParty.take(10).map(_.id.toString).mkString("(e.g., ", ", ", ")")}."
+          )
+      }
+
+      usersWithPrimaryParties.foreach { case (userId, party) =>
+        self ! Subscribe(userId, new PartyState(ApiTypes.Party(party)))
+      }
+
     case UpdatedParties(details) =>
       details.foreach { partyDetails =>
         if (partyDetails.isLocal) {
-          val displayName = partyDetails.displayName match {
-            case Some(value) => value
-            case None => partyDetails.party
-          }
-          self ! Subscribe(
-            displayName,
-            UserConfig(party = ApiTypes.Party(partyDetails.party), role = None, useDatabase = false),
-          )
+          val displayName = partyDetails.displayName.getOrElse(partyDetails.party)
+          self ! Subscribe(displayName, new PartyState(ApiTypes.Party(partyDetails.party)))
         } else {
           log.debug(s"Ignoring non-local party ${partyDetails.party}")
         }
       }
 
-    case Subscribe(displayName, config) =>
+    case Subscribe(displayName, partyState) =>
       if (!state.parties.contains(displayName)) {
-        log.info(s"Starting actor for $displayName")
-        val partyState = new PartyState(config)
-        startPartyActor(state.ledgerClient, partyState)
-        context.become(connected(state.copy(parties = state.parties + (displayName -> partyState))))
+        log.info(s"Starting actor for ${partyState.name} (aka $displayName)")
+
+        // start party actor if needed (since users subscribe to their primary party,
+        // we may subscribe to the same party under different display names, but we should only create one actor per party)
+        val partyActorName = partyState.actorName
+        if (context.child(partyActorName).isEmpty)
+          context.actorOf(
```

Comment:
This is not new behavior, but mentioning it nevertheless: This starts an actor for each discovered party. Each actor uses its own in-memory database to store a complete copy of the whole transaction stream from ledger begin. This means Navigator backend will blow up unless the ledger is "small".

Reply:
Thanks again, this is super helpful!